### PR TITLE
[v6r20] PilotsLogging DB/Service/Client

### PR DESCRIPTION
**Pull request by miloszz**
After few attempts to nightmarish rebase from Hell I decided to make a new branch and cherry-pick valuable commits. Previous branch was a total mess, after few rebases it seemed to have cycles of the same commits over and over again. I felt like still solving the same conflicts. Some files looked like a fusion of Frankenstein's monster and Lady Gaga, with reordered or doubled lines. Lesson learned: PR left alone for too long develops awful mutations.

All comments from previous PR are addressed.

BEGINRELEASENOTES

*WorkloadManagementSystem
NEW: DB, Service and Client for handling extended pilot logging.

ENDRELEASENOTES

### CODE REVIEW

File: WorkloadManagementSystem/Service/WMSAdministratorHandler.py [hunk 1]
```diff
@@ -677,3 +682,32 @@ def export_getPilotStatistics ( attribute, selectDict ):
...
+  types_deletePilots = [ (list, int, long) ]
+  def export_deletePilots( self, pilotIDs ):
+
+    if isinstance( pilotIDs, (int, long ) ):
+      pilotIDs = [pilotIDs, ]
+
+    result = pilotDB.deletePilots( pilotIDs )
+    if not result['OK']:
+      return result
+    result = pilotsLoggingDB.deletePilotsLogging( pilotIDs )
```
Comment (fstagni) [inline WorkloadManagementSystem/Service/WMSAdministratorHandler.py]: Flag here

File: WorkloadManagementSystem/Service/WMSAdministratorHandler.py [hunk 2]
```diff
@@ -677,3 +682,32 @@ def export_getPilotStatistics ( attribute, selectDict ):
...
+##############################################################################
+  types_clearPilots = [ [int, long], [int, long] ]
+  def export_clearPilots( self, interval = 30, aborted_interval = 7 ):
+
+    result = pilotDB.clearPilots( interval, aborted_interval )
+    if not result[ 'OK' ]:
+      return result
+
+    pilotIDs = result[ 'Value' ]
+    result = pilotsLoggingDB.deletePilotsLogging( pilotIDs )
```
Comment (fstagni) [inline WorkloadManagementSystem/Service/WMSAdministratorHandler.py]: Flag here. And also this `result` is not checked if `['OK']`

File: WorkloadManagementSystem/Service/WMSAdministratorHandler.py [hunk 3]
```diff
@@ -47,12 +50,13 @@ def initializeWMSAdministratorHandler( serviceInfo ):
   jobDB = JobDB()
   pilotDB = PilotAgentsDB()
   taskQueueDB = TaskQueueDB()
+  pilotsLoggingDB = PilotsLoggingDB()
```
Comment (fstagni) [inline WorkloadManagementSystem/Service/WMSAdministratorHandler.py]: Flag here

File: WorkloadManagementSystem/Service/WMSAdministratorHandler.py [hunk 4]
```diff
@@ -21,18 +20,22 @@
...
+from DIRAC.Core.Utilities import DErrno
+from datetime import datetime
 
 __RCSID__ = "$Id$"
 
 # This is a global instance of the database classes
 jobDB = False
 pilotDB = False
 taskQueueDB = False
+pilotsLoggingDB = False
```
Comment (fstagni) [inline WorkloadManagementSystem/Service/WMSAdministratorHandler.py]: I'd say it's better to initialize all as `None`

File: WorkloadManagementSystem/Service/WMSAdministratorHandler.py [hunk 5]
```diff
@@ -21,18 +20,22 @@
...
 from DIRAC.WorkloadManagementSystem.DB.PilotAgentsDB import PilotAgentsDB
 from DIRAC.WorkloadManagementSystem.DB.TaskQueueDB import TaskQueueDB
+from DIRAC.WorkloadManagementSystem.DB.PilotsLoggingDB import PilotsLoggingDB
 from DIRAC.WorkloadManagementSystem.Service.WMSUtilities import getPilotLoggingInfo, getWMSPilotOutput, getGridEnv
 from DIRAC.Resources.Computing.ComputingElementFactory import ComputingElementFactory
 import DIRAC.Core.Utilities.Time as Time
 from DIRAC.ConfigurationSystem.Client.Helpers.Registry import getGroupOption, getUsernameForDN
 from DIRAC.ConfigurationSystem.Client.Helpers.Resources import getQueue
+from DIRAC.Core.Utilities import DErrno
+from datetime import datetime
```
Comment (fstagni) [inline WorkloadManagementSystem/Service/WMSAdministratorHandler.py]: Incorrect import order (pylint)

File: WorkloadManagementSystem/DB/PilotAgentsDB.py [hunk 6]
```diff
@@ -200,8 +200,14 @@ def deletePilots( self, pilotIDs, conn = False ):
 
     if type( pilotIDs ) != type( [] ):
       return S_ERROR( 'Input argument is not a List' )
+    
+    for pilotID in pilotIDs:
```
Comment (fstagni) [inline WorkloadManagementSystem/DB/PilotAgentsDB.py]: flag here

File: WorkloadManagementSystem/DB/PilotsLoggingDB.py [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+    except SQLAlchemyError as e:
+      return S_ERROR("Failed to commit: " + e.message)
+      session.rollback( )
+      session.close( )
+
+    return S_OK( )
+
+  ##########################################################################################
+
+class PilotsLogging( Base ):
```
Comment (fstagni) [inline WorkloadManagementSystem/DB/PilotsLoggingDB.py]: Maybe this is just my taste, but shouldn't this be defined at the beginning of the module?

File: WorkloadManagementSystem/Service/JobManagerHandler.py [hunk 8]
```diff
@@ -240,7 +246,23 @@ def __deleteJob( self, jobID ):
     if not result['OK']:
       gLogger.warn( 'Failed to delete job from the TaskQueue' )
 
-    return S_OK()
+    # if it was the last job for the pilot, clear PilotsLogging about it
```
Comment (fstagni) [inline WorkloadManagementSystem/Service/JobManagerHandler.py]: flag here

**Comment by fstagni** [comment]
BTW didn't you have some (nice?) documentation here?

File: WorkloadManagementSystem/DB/PilotAgentsDB.py (hunk 6 above)
Comment (miloszz) [reply WorkloadManagementSystem/DB/PilotAgentsDB.py]: Yes and no. It's pretty awful place, one DB class would need to access another DB. Luckily, all cases seem to be covered on service level (like in WMSAdministratorHandler above, calls to delete pilot, then delete logging). I'm going to remove this part of code.

Comment (fstagni) [reply WorkloadManagementSystem/DB/PilotAgentsDB.py]: yes, good point.

File: WorkloadManagementSystem/Service/JobManagerHandler.py [hunk 9]
```diff
@@ -31,22 +31,31 @@
...
   gJobLoggingDB = JobLoggingDB()
   gtaskQueueDB = TaskQueueDB()
-  gPilotAgentsDB = PilotAgentsDB()
-  gPilotsLoggingDB = PilotsLoggingDB()
+
+  result = gConfig.getValue('/Services/PilotsLogging/Enable')
+  if not result['OK']:
+    return result
+  if result['Value'] == 'Yes':
+    enablePilotsLogging = True
```
Comment (fstagni) [inline WorkloadManagementSystem/Service/JobManagerHandler.py]: `gConfig.getValue` does not return `S_OK/S_ERROR ` structure.

You could also call it like:
```
enablePilotsLogging = gConfig.getValue('/Services/PilotsLogging/Enable', 'False').lower() in ('yes', 'true')
```

File: WorkloadManagementSystem/Service/JobManagerHandler.py [hunk 10]
```diff
@@ -246,21 +255,22 @@ def __deleteJob( self, jobID ):
...
+        res = gPilotAgentsDB.getJobsForPilot( pilot['PilotID'] )
+        if not res['OK']:
+          return res
+        if not res['Value']:  # if list of jobs for pilot is empty, delete pilot and pilotslogging
+          ret = gPilotAgentsDB.deletePilot( pilot['PilotID'] )
+          if not ret['OK']:
+            return ret
+          ret = gPilotsLoggingDB.deletePilotsLogging( pilot['PilotID'] )
+          if not ret['OK']:
+            return ret
```
Comment (fstagni) [inline WorkloadManagementSystem/Service/JobManagerHandler.py]: Don't you want to keep the interaction with `gPilotAgentsDB`?

**Comment by fstagni** [comment]
In https://github.com/DIRACGrid/DIRAC/pull/3360/files you had some documentation that is not yet here. Also, you were using `DErrno`, while here I don't see it.

File: WorkloadManagementSystem/Service/JobManagerHandler.py (hunk 10 above)
Comment (miloszz) [reply WorkloadManagementSystem/Service/JobManagerHandler.py]: I have to use gPilotAgentDB and gPilotsLoggingDB to keep database separation on DB class level. Service handles multiple DB-classes, each DB-class handles just one database.

File: WorkloadManagementSystem/DB/PilotsLoggingDB.py (hunk 7 above)
Comment (miloszz) [reply WorkloadManagementSystem/DB/PilotsLoggingDB.py]: The main class in the file is PilotsLoggingDB (matching file name) that exposes interface for handling database, this one is just a helper/mapper.

**Pull request merged**